### ISSUE TITLE
async builder support

### ISSUE THREAD

**Issue by ColonelThirtyTwo** [joshua-maros/ouroboros#11]
Self referential structures should be able to accept async field builder functions.

For example, it is currently impossible to make a self-referencing structure with an `Arc<futures_util::lock::Mutex<T>>` head and `futures_util::lock::MutexGuard<'this, T>` tail, since acquiring the mutex is an async operation and so can't be done in the closure for constructing the tail.

Should be relatively straightforward to generate code for - constructures become async funcs, have the closure return a future instead of the value directly, then await on those futures.

Bikeshedding stuff:

* How should async vs non-async be specified?
    * Separate `new_async`, `try_new_async`, etc? Possibly only with `self_referencing` attribute?
    * `self_referencing` attribute altering `new`, `try_new`, etc to make them async?
* Enable only with feature flag? I can imagine someone wanting this, but now that async is stable, don't think it buys us anything.
* Safety? I don't think it's unsafe, but weird corner case stuff might prove me wrong.

**Comment by ColonelThirtyTwo** [comment]
Unfortunately not as easy as I'd hope: https://stackoverflow.com/questions/65316753/generic-for-fnonce-that-returns-a-future-with-a-lifetime

**Comment by petrosagg** [comment]
I hit this today. I'm not sure why the macro generates bounds like this one:

```rust
for<'this> FnOnce(&'this <T as Deref>::Target) -> &'this U
```

and not:

```rust
FnOnce(&<T as Deref>::Target) -> &U
```

I did a small experiment to confirm my intuition and I think they are equivalent. Here rustc happily passes the closure from one function to the other: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd0ad4c850caa1895927360ed1f11ede

Here is also the nomicon reference that mention how this expansion happens https://doc.rust-lang.org/nomicon/hrtb.html

@joshua-maros please correct me if this is wildly wrong for some reason 

I'll try to get a PR for async initialization going!

**Comment by joshua-maros** [comment]
@petrosagg I appreciate the research, however removing the lifetime does not work if the function has more than one argument: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e313a31a9730902e43d7922a88b1f3a9

**Comment by petrosagg** [comment]
You're right. Damn! Thanks for the super quick reply! I'll keep exploring the solution space here :)

**Comment by joshua-maros** [comment]
Thanks, I appreciate it!

**Comment by petrosagg** [comment]
The only escape hatch that I see is that because we're stripping lifetimes before passing them to the closure, meaning that they are `'static`, we don't really care what lifetimes the closure sets to its argument and return type so long as they are the same.

And this is expressible in current Rust!

Here is how `fn new()` looks like today compared to a toy `async fn new_async`:
The only difference is in the very last unsafe block that strips the `'a` lifetime.

For reference, this is the original struct:

```rust
#[self_referencing]
struct MyStruct {
    int_data: Box<i32>,
    #[borrows(int_data)]
    int_reference: &'this i32,
}
```

```rust
pub(super) fn new(
    int_data: Box<i32>,
    int_reference_builder: impl for<'this> FnOnce(
        &'this <Box<i32> as Deref>::Target,
    ) -> &'this i32,
) -> Self {
    let mut result = ::core::mem::MaybeUninit::<Self>::uninit();
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_data) as *mut Box<i32>).write(int_data);
    }
    let int_data_illegal_static_reference = unsafe {
        ::ouroboros::macro_help::stable_deref_and_strip_lifetime(
            &((*result.as_ptr()).int_data),
        )
    };
    let int_reference = int_reference_builder(int_data_illegal_static_reference);
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_reference) as *mut &'static i32)
            .write(int_reference);
    }
    unsafe { result.assume_init() }
}

pub(super) async fn new_async<'a, B, F>(
    int_data: Box<i32>,
    int_reference_builder: B,
) -> Self
where
    B: FnOnce(&'a i32) -> F,
    F: Future<Output=&'a i32>
{
    let mut result = ::core::mem::MaybeUninit::<Self>::uninit();
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_data) as *mut Box<i32>).write(int_data);
    }
    let int_data_illegal_static_reference = unsafe {
        ::ouroboros::macro_help::stable_deref_and_strip_lifetime(
            &((*result.as_ptr()).int_data),
        )
    };
    let int_reference = int_reference_builder(int_data_illegal_static_reference).await;

    unsafe {
        // Strip 'a lifetime from the result
        let int_reference: &'static i32 =  &*(int_reference as *const _);
        ((&mut (*result.as_mut_ptr()).int_reference) as *mut &'static i32)
            .write(int_reference);
    }
    unsafe { result.assume_init() }
}
```

Then this kind of usage works:

```rust
#[tokio::main]
async fn main() {
    let my_value = MyStruct::new_async(Box::new(42), |n| {
        async move { n }
    }).await;

    // Prints 42
    println!("{:?}", my_value.borrow_int_data());
}
```

**Comment by joshua-maros** [comment]
@petrosagg  I think that modification would cause a soundness problem:
```rust
#[tokio::main]
async fn main() {
    let external = Box::new(56);
    let external_ref = &external;
    let my_value = MyStruct::new_async(Box::new(42), |n| {
        async move { external_ref }
    }).await;
    drop(external);
    // Soundness problem!
    println!("{:?}", my_value.borrow_int_data());
}
```
(I have not specifically tested this code but I think it compiles.)

The main thing here is that using `for<'a>` requires the closure to work for *every* possible value of `'a`, whereas using it as a template parameter only requires the closure to work for a *specific* value of `'a`.

**Comment by petrosagg** [comment]
I think you're right :( I'll think harder before posting the next comment :D

**Comment by joshua-maros** [comment]
Lol no worries, I ran into this stuff all the time when I first made the crate.

**Comment by petrosagg** [comment]
I think I have a working version! I worked around the limitation of having a second `impl Future` as the result of the builder closure by making it return a `Pin<Box<dyn Future>>` instead. The downside of course is that it requires an allocation but:

1. It's better than not being able to construct these values at all
2. The allocation will be immediately deallocated once initialization is over
3. When the time comes and the compiler supports the desired syntax it will not be a breaking change to switch from `Pin<Box<dyn Future>>` to `impl Future` since the latter is a superset.

Here are the implementations of `new` and `new_async` side by side. The only difference is an await after running the reference builder:

```rust
pub(super) fn new(
    int_data: Box<i32>,
    int_reference_builder: impl for<'this> FnOnce(
        &'this <Box<i32> as Deref>::Target,
    ) -> &'this i32,
) -> Self {
    let mut result = ::core::mem::MaybeUninit::<Self>::uninit();
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_data) as *mut Box<i32>).write(int_data);
    }
    let int_data_illegal_static_reference = unsafe {
        ::ouroboros::macro_help::stable_deref_and_strip_lifetime(
            &((*result.as_ptr()).int_data),
        )
    };
    let int_reference = int_reference_builder(int_data_illegal_static_reference);
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_reference) as *mut &'static i32)
            .write(int_reference);
    }
    unsafe { result.assume_init() }
}

pub(super) async fn new_async(
    int_data: Box<i32>,
    int_reference_builder: impl for<'this> FnOnce(
        &'this <Box<i32> as Deref>::Target,
    ) -> Pin<Box<dyn Future<Output=&'this i32> + 'this>>,
) -> Self {
    let mut result = ::core::mem::MaybeUninit::<Self>::uninit();
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_data) as *mut Box<i32>).write(int_data);
    }
    let int_data_illegal_static_reference = unsafe {
        ::ouroboros::macro_help::stable_deref_and_strip_lifetime(
            &((*result.as_ptr()).int_data),
        )
    };
    let int_reference = int_reference_builder(int_data_illegal_static_reference).await;
    //                   only difference here -------------------------------------^
    unsafe {
        ((&mut (*result.as_mut_ptr()).int_reference) as *mut &'static i32)
            .write(int_reference);
    }
    unsafe { result.assume_init() }
}
```

And here is an example use:

```rust
#[tokio::main]
async fn main() {
    let my_value = MyStruct::new_async(Box::new(42i32), |n| {
        Box::pin(async move { n })
    }).await;

    // Prints 42
    println!("{:?}", my_value.borrow_int_reference());
}
```

What do you think? If it looks good to you I can wrap it up in a PR :)

**Comment by joshua-maros** [comment]
That looks fantastic! I'll review your PR to see if there are any small code style things to change but other than that it looks great!

**Comment by joshua-maros** [comment]
@ColonelThirtyTwo @petrosagg This is now available in version `0.8.1`.

**Comment by petrosagg** [comment]
Thank you for the quick release!
